### PR TITLE
DCL Parser with AST Object Encapsulation

### PR DESCRIPTION
**Pull request by JD-Howard**
### Objective
One of our big obstacles for truly finishing work with any particular provider has been related to DCL technical debt. This PR adds proper DCL Parsing that fully wraps the document in AST Objects/Interfaces. By itself, it doesn't actually add any new functionality for end users, but is the foundation for various provider enhancements.

### Abstractions
This was designed for future reliance on the `IDclFragment` and its derived `IDclContainer` interfaces, but there is no actual inheritance between the concrete classes. All DCL AST Objects are fragments, but 2 of the 3 are containers. Both interfaces are defined in the `DclInterfaces.ts` file and all the concrete objects implement functionality ordered by IDclFragement->IDclContainer->Class

**Concrete Classes**
`DclAtom` represents single/block comments, strings and generally similar to the LispAtom that holds sequential characters
`DclAttribute` Single line sequence of DclAtoms that are not comments and between `{` and `}` brackets
`DclTile` Handles standard and dialog definition tiles, contains attributes, other tiles, comments & initiated by an `{` bracket

The DCL parser has been tested for malformed syntax and is performing in such a way that it doesn't lose any elements, provides predictable context that will support providers and in the absence of trailing spaces/tabs can read and exactly reproduce a document.

The ReadOnlyDocument has been updated to provide the containerized view of a DCL document in the same way an LSP file would. LSP's used `documentContainer` property and now has a `documentDclContainer` property for DCL files.

Also fixed argument bugs in previous tests.

### Tests performed
All new functionality covered by this PR is between 95% and 100% on the code coverage; with an above 98% average. The tests could be divided up more than they currently are, but a large effort was made to ensure basic and edge cases all perform exactly as expected. All new tests have been performed on Mac and are working as expected.

### Screen shot
Nothing to show because this is an analytical support PR

### CODE REVIEW

**Comment by JD-Howard** [comment]
@Sen-real 
This should be pretty solid, but please give it a review when you get a chance. I'll get it tested on my Mac later today.

File: extension/src/test/suite/astObjects.DclAttribute.test.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+			expect(sut1.lastAtom.symbol).to.equal(';');
+			expect(sut2.lastAtom.symbol).to.equal(';');
+		}
+		catch (err) {
+			assert.fail("The expected container and/or value was not returned");
+		}
+	});
+
+
+	test("DclAttribute Positional Properties", function () {
```
Comment (JD-Howard) [inline extension/src/test/suite/astObjects.DclAttribute.test.ts:150]: I need to change this description, it is clearly a copy of the previous test.

File: extension/src/test/suite/parsing.dclParser.test.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	const good1 = 'myDCL : dialog {\n\tok_cancel;\n\t}';
+	const good2 = 'myDCL : dialog {\n\t: button {\n\t\twidth = 100;\n\t\tlabel = "random";\n\t\t}\n\tok_cancel;\n\t}';
+	const poor0 = 'myDCL : dialog {\n\tok_cancel;';
+	const poor1 = 'myDCL : dialog {:button{} ok_cancel;';
+	const poor2 = 'myDCL : dialog {\n\t: button {\n\t\twidth = 100;\n\t\tlabel = "random";\n\t\tok_cancel;\n\t}';
+	const poor3 = 'myDCL : dialog {\n\t: button {\n\t\twidth = 100\t\t\tlabel = "random";\n\t\t}\n\tok_cancel;\n\t}';
+	const poor4 = 'myDCL : dialog {\n\t: button {\n\t\twidth = 100;\n\t\tlabel "random";\n\t\t}\n\tok_cancel;';
+	const poor5 = 'myDCL : dialog {\n\t: button {\n\t\t= 100;\n\t\tlabel = "random"\n\tok_cancel;\n\t}';
+
+	// TODO: Add malformed syntax tests
```
Comment (JD-Howard) [inline extension/src/test/suite/parsing.dclParser.test.ts:72]: I did this and I need to delete this comment

File: extension/src/parsing/dclParser.ts [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+		const next = ctx.data.charCodeAt(ctx.dataIndex + 1); 
+
+		if (curr === openBracket) {	// tile/attribute breakpoint
+			ctx.saveTempDataAndCurrentCharToFragments()
+			   .moveNext()
+			   .createNewTileScopeFromFragments();
+		} else if (curr === closeBracket) { // tile/attribute breakpoint
+			ctx.saveTempData()
+			   .saveFragmentsAsAttribute()
+			   .saveTempDataAndCurrentCharToFragments()
```
Comment (JD-Howard) [inline extension/src/parsing/dclParser.ts:232]: This looks sketchy, but is doing what is intended. If you want, I can go create a `saveCurrentCharToFragments` function so this doesn't look redundant.

File: extension/src/parsing/dclParser.ts [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+			   .saveFragmentsAsAttribute();
+		} else if (curr === forwardSlash) { // attribute breakpoint
+			ctx.saveTempData()
+			   .saveFragmentsAsAttribute()
+			   .startGroupPointer()
+			   .processCommentAndSaveAsFragment(next === asterisk);
+		} else if (curr === cr || curr === lf) { // fragment/attribute breakpoint
+			ctx.saveTempData()
+			   .saveFragmentsAsAttribute()
+			   .incrementAtLineBreak(curr) // this may need to account for the subsequent move next?????????
```
Comment (JD-Howard) [inline extension/src/parsing/dclParser.ts:247]: I need to delete this comment, it has subsequently been proven to be working as intended as is.

File: extension/src/parsing/dclParser.ts [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+const equalSign = '='.charCodeAt(0);
+const dblQuote = '"'.charCodeAt(0);
+const forwardSlash = '/'.charCodeAt(0);
+const backSlash = '\\'.charCodeAt(0);
+const asterisk = '\*'.charCodeAt(0);
+
+
+export function getDocumentTileContainer(DocumentOrContent: vscode.TextDocument|string): DclTile {
+	const ctx = new DclDocContext(DocumentOrContent);
+	_getDclDocumentContainer(ctx);
```
Comment (JD-Howard) [inline extension/src/parsing/dclParser.ts:24]: should probably move this into the try block

**Comment by JD-Howard** [comment]
@Sen-real 
I've resolved all the issues I've found and anything having to do with CodeQL. I also now tested it on my Mac and the only tests failing are the Lisp formatting tests that have always failed on Mac.

**Comment by Sen-real** [comment]
> @Sen-real I've resolved all the issues I've found and anything having to do with CodeQL. I also now tested it on my Mac and the only tests failing are the Lisp formatting tests that have always failed on Mac.

Thank you @JD-Howard , I'm looking at this PR and will have a try locally.

File: extension/src/astObjects/dclTile.ts [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    get firstNonComment(): IDclFragment {
+        for (let i = 0; i < this.length; i++) {
+            const item = this.atoms[i];
+            if (item.isComment){
+                continue;
+            }
+            if (item.asContainer) {
+                return null;
```
Comment (JD-Howard) [inline extension/src/astObjects/dclTile.ts:84]: This rejects anything that that is a container because any **non-root** tile should always have at least ": tiletype {" as prefix DclAtoms. If this is confusing, I can add a comment.

**Review by Sen-real — APPROVED** [review]
